### PR TITLE
fix(issue): Branch-mode milestone entry fails on dirty working tree (no stash guard)

### LINKED ISSUE #5927: Branch-mode milestone entry fails on dirty working tree (no stash guard)
## Problem

Auto-mode branch-isolation milestone entry fails when the working tree has uncommitted changes. `git checkout milestone/M019` aborts with "Your local changes would be overwritten by checkout", and auto-mode surfaces the error and stops. No stash/restore logic exists on the entry path — only on the merge path.

## Root Cause

`enterBranchModeForMilestone` in `auto-worktree.js:764-814` calls `nativeCheckoutBranch(basePath, branch)` at **line 813** without any dirty-working-tree check or stash. The caller `enterMilestone` in `worktree-lifecycle.js:331-333` also has no stash/clean logic before invoking it.

The existing `preflightCleanRoot` function (in `clean-root-preflight.js:56`) does exactly what's needed — stash dirty state, proceed, restore — but it's only wired into the **merge** path (`auto/phases.js:305`), not the **entry** path.

```js
// auto-worktree.js ~line 813 — no stash guard before checkout
const checkoutResult = await nativeCheckoutBranch(basePath, branch);
```

**Two affected code paths:**
1. `worktree-lifecycle.js:333` — branch-mode milestone entry (the crash here)
2. `auto-start.js:831` — `isolation:none` stale-branch recovery checkout (same pattern, same missing stash)

## Expected Behavior

Before calling `nativeCheckoutBranch` in `enterBranchModeForMilestone`, the code should:
1. Check for dirty working tree (`git status --porcelain`)
2. If dirty, `git stash push` (like `preflightCleanRoot` does)
3. Perform the checkout
4. `git stash pop` on success, or restore in the catch block on failure

The same stash-push/pop pattern should be applied to the `auto-start.js:831` isolation:none recovery checkout.

## Environment
- GSD version: 2.82.0
- Model: claude-opus-4-6
- Unit: plan-slice M019/S01 (auto-mode was entering M019 when it failed)

## Reproduction Context

- Branch isolation mode configured in `.gsd/PREFERENCES.md`
- Previous session added `.dropboxignore` template entries to `.gitignore` but didn't commit them
- Auto-mode completed M018, then tried to enter M019
- `git checkout milestone/M019` failed because `.gitignore` had uncommitted changes
- Auto-mode surfaced the error and stopped

## Forensic Evidence

**Error message:**
```
Command failed: git checkout milestone/M019
error: Your local changes to the following files would be overwritten by checkout:
  .gitignore
Please commit your changes or stash them before you switch branches.
Aborting
```

**Journal trace (last events):**
- `worktree-skip` → `dispatch-stop` → `dispatch-match rule=refining → refine-slice M019/S01` → `artifact-verification-retry` → `auto-exit`
- 26 worktree orphans (in-progress-unmerged), 21 (complete-unmerged)

**Related but distinct issues:**
- #5067 covers `preflightCleanRoot` missing Windows db-cycle before stash push during *merge*
- #3141 covers hardening git operations in the merge cycle
- Neither covers the *entry* path, which has zero stash logic

---
*Auto-generated by `/gsd forensics`*

### PR DESCRIPTION
**Pull request by jeremymcs**
## Summary
- Added stash-protected branch checkout for milestone entry and isolation:none recovery, with focused regression tests passing.

## Verification
- Completed in the repository worktree before push.

## Related Issue
- Closes #5927
- [#5927 Branch-mode milestone entry fails on dirty working tree (no stash guard)](https://github.com/gsd-build/gsd-2/issues/5927)

## Repo
- `gsd-build/gsd-2`

## Branch
- `issue/5927-branch-mode-milestone-entry-fails-on-dir-1778725001`

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved branch switching reliability with enhanced safeguards that preserve working directory state during checkout operations and provide better recovery mechanisms when issues occur.

* **Tests**
  * Added comprehensive test coverage for branch checkout safety and working directory state preservation.

[Review Change Stack](https://app.coderabbit.ai/change-stack/gsd-build/gsd-2/pull/5948)

### CODE REVIEW

File: src/resources/extensions/gsd/auto-worktree.ts [hunk 1]
```diff
@@ -1007,7 +1007,45 @@ export function enterBranchModeForMilestone(
...
+  } catch (err) {
+    if (stashed) {
+      try {
+        popStashByRef(basePath, stashMarker);
+      } catch (restoreErr) {
+        logWarning("worktree", `git stash pop failed during checkout restore: ${restoreErr instanceof Error ? restoreErr.message : String(restoreErr)}`);
+      }
+    }
+    throw err;
+  }
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/auto-worktree.ts]: ⏳ **Accepted** — this comment requires code changes. Queuing fix...
🧰 **Agent running** — `claude` is working on the fix...
✅ **Agent completed** — verifying changes...
🎉 **Resolved** — addressed in commit `9331bbd`.
🎉 **Resolved** — addressed in commit `97ad60c`.

File: src/resources/extensions/gsd/tests/checkout-branch-stash-guard.test.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    assert.throws(
+      () => checkoutBranchWithStashGuard(repo, "milestone/DOES-NOT-EXIST", "test-failure"),
+    );
+
+    const status = git(["status", "--porcelain"], repo);
+    assert.match(status, /note\.txt/);
+    const stashList = git(["stash", "list"], repo).trim();
+    assert.equal(stashList, "");
+  });
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/tests/checkout-branch-stash-guard.test.ts:63]: ⏳ **Accepted** — this comment requires code changes. Queuing fix...
🧰 **Agent running** — `claude` is working on the fix...
✅ **Agent completed** — verifying changes...
🎉 **Resolved** — addressed in commit `9331bbd`.
🎉 **Resolved** — addressed in commit `97ad60c`.

**Comment by jeremymcs** [comment]
🤖 Dispatched `claude` with the following prompt:

<details>
<summary>Agent prompt (click to expand)</summary>

````text
You are acting as an autonomous PR babysitter for gsd-build/gsd-2 PR #5948.
PR URL: https://github.com/gsd-build/gsd-2/pull/5948
Base repository: gsd-build/gsd-2
Head repository: gsd-build/gsd-2
Head branch: issue/5927-branch-mode-milestone-entry-fails-on-dir-1778725001
Head remote: origin
You are running inside an isolated app-owned worktree under ~/.patchdeck.
Make only targeted changes that resolve the approved tasks.
Do not wait for user input, confirmation, or approval at any point.
Do not rewrite unrelated files.
Use the available git tooling for inspection and verification only.
If dependencies are missing, install them using the repository's lockfile/package manager as needed inside this isolated worktree.
Leave file edits uncommitted; the babysitter app will handle Git finalization after your run.
GitHub follow-up replies and review-thread resolution will be handled by the babysitter after your run.
If a task is invalid after inspection, explain it in your final response and include the exact audit token.

Approved review-comment tasks:
1. [review_comment] coderabbitai[bot]
   file=src/resources/extensions/gsd/auto-worktree.ts line=1048
   sourceId=3238679770 sourceUrl=https://github.com/gsd-build/gsd-2/pull/5948#discussion_r3238679770
   replyKind=review_thread threadId=PRRT_kwDORj5hb86B8a23 threadResolved=false
   auditToken=codefactory-feedback:gh-review-comment-3238679770
   _⚠️ Potential issue_ | _🟡 Minor_ | _⚡ Quick win_

**Clarify error semantics when checkout succeeds but stash restore fails.**

If `nativeCheckoutBranch` (line 1036) succeeds but `popStashByRef` (line 1038) throws, the catch block (line 1042) attempts to pop the stash again, then rethrows the original pop error. This means:

- The branch switch **succeeded** (HEAD moved to the target branch).
- The working tree changes could not be restored (stash entry remains in the stash list).
- The function throws an error, even though the primary operation (checkout) completed.

This could confuse callers: an error is surfaced even though the branch was switched. The user might retry the operation, not realizing HEAD already moved.

**Recommendation**: Add a comment explaining this edge case, or consider detecting whether the checkout succeeded before deciding to throw. For example, you could wrap the checkout in a separate try-catch, or verify HEAD after the pop failure to emit a more descriptive error message.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@src/resources/extensions/gsd/auto-worktree.ts` around lines 1036 - 1048, The
current catch rethrows the original error when popStashByRef fails even if
nativeCheckoutBranch already succeeded, which can mislead callers; update the
logic around nativeCheckoutBranch and popStashByRef (and the stashMarker
handling) to detect whether the checkout actually succeeded before deciding to
throw: either separate the checkout into its own try/catch and only treat
stash-pop failures as non-fatal (log a warning and throw a new error that
clearly states "checkout succeeded but stash restore failed" or return a result
indicating partial success), or after a pop failure verify HEAD (e.g., check
current branch) and then rethrow or return a descriptive error that notes the
branch switch succeeded but the stash entry was not restored; also add a short
inline comment near nativeCheckoutBranch/popStashByRef explaining this edge case
so future readers understand the semantics.
```

</details>

<!-- fingerprinting:phantom:poseidon:puma -->

<!-- This is an auto-generated comment by CodeRabbit -->
2. [review_comment] coderabbitai[bot]
   file=src/resources/extensions/gsd/tests/checkout-branch-stash-guard.test.ts line=61
   sourceId=3238679774 sourceUrl=https://github.com/gsd-build/gsd-2/pull/5948#discussion_r3238679774
   replyKind=review_thread threadId=PRRT_kwDORj5hb86B8a27 threadResolved=false
   auditToken=codefactory-feedback:gh-review-comment-3238679774
   _🛠️ Refactor suggestion_ | _🟠 Major_ | _⚡ Quick win_

**Add test coverage for checkout-success + pop-failure scenario.**

The current tests cover:
1. Successful checkout with clean stash restore (lines 32-47)
2. Failed checkout with successful stash restore (lines 49-61)

Missing: checkout succeeds but stash pop fails. This edge case was identified in the implementation review—if the target branch has conflicting changes, the stash pop after a successful checkout can throw. The function would rethrow the pop error even though HEAD already moved to the target branch.

To test this scenario, you could:
1. Create branch A with `file.txt = "A"`
2. Create branch B with `file.txt = "B"`
3. Checkout A, modify `file.txt` to "local"
4. Call `checkoutBranchWithStashGuard(repo, "B", ...)`
5. Verify that the function throws (pop conflict)
6. Verify that HEAD is on branch B (checkout succeeded)
7. Verify that a stash entry remains (pop failed, so stash wasn't consumed)

This would confirm the error-handling contract and document the expected behavior for this edge case.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@src/resources/extensions/gsd/tests/checkout-branch-stash-guard.test.ts`
around lines 49 - 61, Add a new test in checkout-branch-stash-guard.test that
exercises the "checkout succeeds but stash pop fails" edge case: create two
branches (A and B) with differing contents for a tracked file, commit B and A
appropriately, check out A, make a local dirty change to the file, then call
checkoutBranchWithStashGuard(repo, "B", "test-pop-failure") and assert it
throws; after the call assert HEAD is on branch "B" (checkout succeeded) and
that git(["stash", "list"], repo) still contains a stash entry (pop failed and
stash not consumed), using existing helpers like createRepo, git, writeFileSync
and reusing the test pattern from other cases to set up commits and assertions.
```

</details>

<!-- fingerprinting:phantom:poseidon:puma -->

<!-- This is an auto-generated comment by CodeRabbit -->

Approved status-check tasks:
None

Approved documentation tasks:
None

When done:
1) Run the relevant verification for your changes.
2) Leave any changed files in the worktree for the babysitter app to finalize.
3) For each feedback item you addressed or were blocked on, emit a summary block in the following format:
   FEEDBACK_SUMMARY_START <auditToken>
   <A concise 1-2 sentence summary of what you did or why you were blocked>
   FEEDBACK_SUMMARY_END
   Include one block per audit token. These summaries will be posted as follow-up comments on the PR.
4) If documentation tasks were assigned, emit exactly one docs summary block in the following format:
   DOCS_SUMMARY_START <changed|no_change>
   <A concise 1-2 sentence summary of the docs you updated, or why no docs changes were necessary after inspection>
   DOCS_SUMMARY_END
````

</details>

File: src/resources/extensions/gsd/auto-worktree.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/auto-worktree.ts]: Addressed in commit `9331bbd` by the latest babysitter run.

Split `checkoutBranchWithStashGuard` so checkout and stash-pop are in separate try blocks. A post-checkout pop failure now throws a distinct wrapped error stating the branch switch succeeded but the stash entry remains, with an inline comment documenting the two failure modes.

File: src/resources/extensions/gsd/tests/checkout-branch-stash-guard.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/tests/checkout-branch-stash-guard.test.ts:63]: Addressed in commit `9331bbd` by the latest babysitter run.

Added a third test case that creates a divergent branch B, dirties the file on main, calls the guard, and asserts the function throws the new "checkout succeeded but stash restore failed" error while HEAD is on B and the stash entry remains in `git stash list`.

**Comment by jeremymcs** [comment]
⏳ **Accepted** — this comment requires code changes. Queuing fix...
🧰 **Agent running** — `codex` is working on the fix...
✅ **Agent completed** — verifying changes...

File: src/resources/extensions/gsd/tests/checkout-branch-stash-guard.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/tests/checkout-branch-stash-guard.test.ts:63]: ⏳ **Accepted** — this comment requires code changes. Queuing fix...
🧰 **Agent running** — `codex` is working on the fix...
✅ **Agent completed** — verifying changes...

File: src/resources/extensions/gsd/auto-worktree.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/auto-worktree.ts]: ⏳ **Accepted** — this comment requires code changes. Queuing fix...
🧰 **Agent running** — `codex` is working on the fix...
✅ **Agent completed** — verifying changes...

**Comment by jeremymcs** [comment]
🤖 Dispatched `codex` with the following prompt:

<details>
<summary>Agent prompt (click to expand)</summary>

````text
You are acting as an autonomous PR babysitter for gsd-build/gsd-2 PR #5948.
PR URL: https://github.com/gsd-build/gsd-2/pull/5948
Base repository: gsd-build/gsd-2
Head repository: gsd-build/gsd-2
Head branch: issue/5927-branch-mode-milestone-entry-fails-on-dir-1778725001
Head remote: origin
You are running inside an isolated app-owned worktree under ~/.patchdeck.
Make only targeted changes that resolve the approved tasks.
Do not wait for user input, confirmation, or approval at any point.
Do not rewrite unrelated files.
Use the available git tooling for inspection and verification only.
If dependencies are missing, install them using the repository's lockfile/package manager as needed inside this isolated worktree.
Leave file edits uncommitted; the babysitter app will handle Git finalization after your run.
GitHub follow-up replies and review-thread resolution will be handled by the babysitter after your run.
If a task is invalid after inspection, explain it in your final response and include the exact audit token.

Approved review-comment tasks:
1. [review] coderabbitai[bot]
   file=n/a line=n/a
   sourceId=4286812118 sourceUrl=https://github.com/gsd-build/gsd-2/pull/5948#pullrequestreview-4286812118
   replyKind=review threadId=n/a threadResolved=n/a
   auditToken=codefactory-feedback:gh-review-4286812118
   **Actionable comments posted: 2**

<details>
<summary>🧹 Nitpick comments (2)</summary><blockquote>

<details>
<summary>src/resources/extensions/gsd/tests/checkout-branch-stash-guard.test.ts (1)</summary><blockquote>

`32-47`: _💤 Low value_

**Consider strengthening the working-tree restoration assertion.**

The test verifies that `note.txt` appears in `git status --porcelain` output (line 46), but doesn't verify that the content is exactly `"dirty\n"` as written on line 37. You could use `readFileSync` to assert the exact restored content, which would provide stronger coverage that the stash/restore cycle preserves file contents correctly.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@src/resources/extensions/gsd/tests/checkout-branch-stash-guard.test.ts`
around lines 32 - 47, In the test "restores dirty working tree after successful
checkout" strengthen the assertion by reading the working-tree file directly
after checkoutBranchWithStashGuard and asserting its content equals "dirty\n";
specifically, after calling checkoutBranchWithStashGuard(repo, "milestone/M001",
"test-success") use readFileSync(join(repo, "note.txt"), "utf8") and
assert.strictEqual (or assert.equal) that the returned string is "dirty\n" in
addition to the existing git status check, so the stash/restore cycle is
verified to preserve file contents.
```

</details>

</blockquote></details>
<details>
<summary>src/resources/extensions/gsd/auto-worktree.ts (1)</summary><blockquote>

`1033-1033`: _💤 Low value_

**Consider a more robust stash-detection mechanism.**

The current implementation infers whether a stash entry was created by checking if `git stash push` output includes "No local changes to save". This is fragile—git could change the message text or localize output. A more defensive approach would be to compare `git stash list` before and after the push, or to check `git status --porcelain` after the push completes.

That said, this pattern is consistent with the existing pre-merge stash logic at line 2026, so it's not introducing new risk.

<details>
<summary>🤖 Prompt for AI Agents</summary>

```
Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@src/resources/extensions/gsd/auto-worktree.ts` at line 1033, The current
assignment to the variable `stashed` uses a fragile string-check (stashed =
!output.includes("No local changes to save")) after `git stash push`; replace
that with a robust detection: capture the output of `git stash list` (or its
hash entries) immediately before the `git stash push` call and again after, and
set `stashed` based on whether the list changed (or alternatively run `git
status --porcelain` after the push and treat a clean/unchanged working tree as
stashed=false); update the code path around the `git stash push` invocation so
`stashed` is derived from list-diff or porcelain-check rather than a localized
output string.
```

</details>

</blockquote></details>

</blockquote></details>

<details>
<summary>🤖 Prompt for all review comments with AI agents</summary>

```
Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

Inline comments:
In `@src/resources/extensions/gsd/auto-worktree.ts`:
- Around line 1036-1048: The current catch rethrows the original error when
popStashByRef fails even if nativeCheckoutBranch already succeeded, which can
mislead callers; update the logic around nativeCheckoutBranch and popStashByRef
(and the stashMarker handling) to detect whether the checkout actually succeeded
before deciding to throw: either separate the checkout into its own try/catch
and only treat stash-pop failures as non-fatal (log a warning and throw a new
error that clearly states "checkout succeeded but stash restore failed" or
return a result indicating partial success), or after a pop failure verify HEAD
(e.g., check current branch) and then rethrow or return a descriptive error that
notes the branch switch succeeded but the stash entry was not restored; also add
a short inline comment near nativeCheckoutBranch/popStashByRef explaining this
edge case so future readers understand the semantics.

In `@src/resources/extensions/gsd/tests/checkout-branch-stash-guard.test.ts`:
- Around line 49-61: Add a new test in checkout-branch-stash-guard.test that
exercises the "checkout succeeds but stash pop fails" edge case: create two
branches (A and B) with differing contents for a tracked file, commit B and A
appropriately, check out A, make a local dirty change to the file, then call
checkoutBranchWithStashGuard(repo, "B", "test-pop-failure") and assert it
throws; after the call assert HEAD is on branch "B" (checkout succeeded) and
that git(["stash", "list"], repo) still contains a stash entry (pop failed and
stash not consumed), using existing helpers like createRepo, git, writeFileSync
and reusing the test pattern from other cases to set up commits and assertions.

---

Nitpick comments:
In `@src/resources/extensions/gsd/auto-worktree.ts`:
- Line 1033: The current assignment to the variable `stashed` uses a fragile
string-check (stashed = !output.includes("No local changes to save")) after `git
stash push`; replace that with a robust detection: capture the output of `git
stash list` (or its hash entries) immediately before the `git stash push` call
and again after, and set `stashed` based on whether the list changed (or
alternatively run `git status --porcelain` after the push and treat a
clean/unchanged working tree as stashed=false); update the code path around the
`git stash push` invocation so `stashed` is derived from list-diff or
porcelain-check rather than a localized output string.

In `@src/resources/extensions/gsd/tests/checkout-branch-stash-guard.test.ts`:
- Around line 32-47: In the test "restores dirty working tree after successful
checkout" strengthen the assertion by reading the working-tree file directly
after checkoutBranchWithStashGuard and asserting its content equals "dirty\n";
specifically, after calling checkoutBranchWithStashGuard(repo, "milestone/M001",
"test-success") use readFileSync(join(repo, "note.txt"), "utf8") and
assert.strictEqual (or assert.equal) that the returned string is "dirty\n" in
addition to the existing git status check, so the stash/restore cycle is
verified to preserve file contents.
```

</details>

<details>
<summary>🪄 Autofix (Beta)</summary>

Fix all unresolved CodeRabbit comments on this PR:

- [ ] <!-- {"checkboxId": "4b0d0e0a-96d7-4f10-b296-3a18ea78f0b9"} --> Push a commit to this branch (recommended)
- [ ] <!-- {"checkboxId": "ff5b1114-7d8c-49e6-8ac1-43f82af23a33"} --> Create a new PR with the fixes

</details>

---

<details>
<summary>ℹ️ Review info</summary>

<details>
<summary>⚙️ Run configuration</summary>

**Configuration used**: defaults

**Review profile**: CHILL

**Plan**: Pro Plus

**Run ID**: `d45d5d1c-ad03-4de2-a5ac-bba5546d317f`

</details>

<details>
<summary>📥 Commits</summary>

Reviewing files that changed from the base of the PR and between 3709f22a50338754d92d1d217b7b95a4edca7223 and 070fc40bdccfad70b95b26ce754fc888058deb5c.

</details>

<details>
<summary>📒 Files selected for processing (3)</summary>

* `src/resources/extensions/gsd/auto-start.ts`
* `src/resources/extensions/gsd/auto-worktree.ts`
* `src/resources/extensions/gsd/tests/checkout-branch-stash-guard.test.ts`

</details>

</details>

<!-- This is an auto-generated comment by CodeRabbit for review status -->
2. [review_comment] coderabbitai[bot]
   file=src/resources/extensions/gsd/tests/checkout-branch-stash-guard.test.ts line=61
   sourceId=3238688129 sourceUrl=https://github.com/gsd-build/gsd-2/pull/5948#discussion_r3238688129
   replyKind=review_thread threadId=PRRT_kwDORj5hb86B8a27 threadResolved=true
   auditToken=codefactory-feedback:gh-review-comment-3238688129
   `@jeremymcs`, here's the plan to implement the missing edge-case test.



- Add a third test case to the existing suite covering "checkout succeeds but stash pop fails"
- Set up two branches with conflicting content on the same tracked file
- Dirty the working tree with a local modification that will conflict on pop
- Assert the function throws, HEAD is on the target branch, and a stash entry remains

<!-- <requirement>
[requirement add_pop_failure_test_1/1]
Add a new test `"leaves HEAD on target branch and stash intact when pop conflicts"` to `src/resources/extensions/gsd/tests/checkout-branch-stash-guard.test.ts` after line 61, using the existing helpers.

The setup:
1. `createRepo` starts with `main` having `note.txt = "base\n"`.
2. Create branch `milestone/B` and commit `note.txt = "branch-b\n"` to it.
3. Switch back to `main`.
4. Dirty `note.txt` with `"local-change\n"` — this will conflict when popped onto `milestone/B` because both sides modified `note.txt`.
5. Call `checkoutBranchWithStashGuard(repo, "milestone/B", "test-pop-failure")`.
6. Assert the call `throws`.
7. Assert `branch --show-current` returns `"milestone/B"` (checkout succeeded before pop).
8. Assert `stash list` is non-empty (pop failed, stash entry was not consumed).

File: `src/resources/extensions/gsd/tests/checkout-branch-stash-guard.test.ts`

```diff
   test("restores dirty working tree when checkout throws", (t) => {
     const repo = createRepo(t);
     writeFileSync(join(repo, "note.txt"), "dirty\n");

     assert.throws(
       () => checkoutBranchWithStashGuard(repo, "milestone/DOES-NOT-EXIST", "test-failure"),
     );

     const status = git(["status", "--porcelain"], repo);
     assert.match(status, /note\.txt/);
     const stashList = git(["stash", "list"], repo).trim();
     assert.equal(stashList, "");
   });
+
+  test("leaves HEAD on target branch and stash intact when pop conflicts", (t) => {
+    const repo = createRepo(t);
+
+    // Create milestone/B with different content for note.txt
+    git(["checkout", "-b", "milestone/B"], repo);
+    writeFileSync(join(repo, "note.txt"), "branch-b\n");
+    git(["add", "note.txt"], repo);
+    git(["commit", "-m", "branch-b content"], repo);
+
+    // Return to main and dirty the same file so the pop will conflict
+    git(["checkout", "main"], repo);
+    writeFileSync(join(repo, "note.txt"), "local-change\n");
+
+    // The checkout will succeed (HEAD moves to milestone/B) but the stash pop
+    // will conflict because both sides modified note.txt — expect a throw.
+    assert.throws(
+      () => checkoutBranchWithStashGuard(repo, "milestone/B", "test-pop-failure"),
+    );
+
+    // Checkout succeeded even though pop failed
+    const branch = git(["branch", "--show-current"], repo).trim();
+    assert.equal(branch, "milestone/B");
+
+    // Stash entry is still present because pop was not consumed cleanly
+    const stashList = git(["stash", "list"], repo).trim();
+    assert.ok(stashList.length > 0, "expected a stash entry to remain after failed pop");
+  });
```
</requirement> -->

---

Failed to handle agent chat message. Please try again.



<!-- This is an auto-generated reply by CodeRabbit -->
3. [review_comment] coderabbitai[bot]
   file=src/resources/extensions/gsd/auto-worktree.ts line=1048
   sourceId=3238692408 sourceUrl=https://github.com/gsd-build/gsd-2/pull/5948#discussion_r3238692408
   replyKind=review_thread threadId=PRRT_kwDORj5hb86B8a23 threadResolved=true
   auditToken=codefactory-feedback:gh-review-comment-3238692408
   `@jeremymcs`, thank you for accepting — here's the planned fix:



## Fix: Clarify error semantics when checkout succeeds but stash restore fails

- Split the single `try/catch` into two phases: **checkout phase** and **stash-pop phase**
- Stash-pop failure *after* a successful checkout is non-fatal: log a warning and throw a descriptive error clearly stating the checkout succeeded but the stash was not restored (so callers/users don't retry a branch switch already done)
- Stash-pop failure *after* a failed checkout remains the current behavior: attempt restore, swallow pop error, rethrow the checkout error
- Add an inline comment explaining the edge case

<!-- <requirement>
[requirement fix_checkout_stash_guard_error_semantics_1/1]
Restructure `checkoutBranchWithStashGuard` in `src/resources/extensions/gsd/auto-worktree.ts` so that:

1. The checkout and stash-pop phases are separated.
2. If checkout **fails**: attempt stash restore, swallow pop errors, rethrow the original checkout error (unchanged behavior).
3. If checkout **succeeds** but stash-pop **fails**: log a warning and throw a new descriptive error clearly stating the branch switch completed but the stash was not restored — do NOT retry the pop from the outer catch.
4. Add an inline comment near the stash-pop call explaining the edge case.

```diff
 export function checkoutBranchWithStashGuard(
   basePath: string,
   branch: string,
   reason: string,
 ): void {
   let stashMarker: string | null = null;
   let stashed = false;
+  let checkoutSucceeded = false;
   try {
     const status = nativeWorkingTreeStatus(basePath).trim();
     if (status.length > 0) {
       stashMarker = `gsd-checkout-stash:${reason}:${process.pid}:${Date.now()}:${process.hrtime.bigint().toString(36)}`;
       const output = execFileSync(
         "git",
         ["stash", "push", "--include-untracked", "-m", `gsd: checkout stash [${stashMarker}]`],
         {
           cwd: basePath,
           stdio: ["ignore", "pipe", "pipe"],
           encoding: "utf-8",
         },
       );
       stashed = !output.includes("No local changes to save");
     }
 
     nativeCheckoutBranch(basePath, branch);
+    checkoutSucceeded = true;
 
-    if (stashed) popStashByRef(basePath, stashMarker);
+    // Stash restore after a successful checkout. If this throws, the branch
+    // switch already happened (HEAD moved). Treat as a non-fatal partial
+    // failure: warn and surface a descriptive error so callers know the
+    // checkout completed — retrying would be a no-op or switch HEAD again.
+    if (stashed) {
+      try {
+        popStashByRef(basePath, stashMarker);
+      } catch (popErr) {
+        const msg = popErr instanceof Error ? popErr.message : String(popErr);
+        logWarning(
+          "worktree",
+          `Branch switched to "${branch}" but stash restore failed — stash entry preserved for manual recovery: ${msg}`,
+        );
+        throw new GSDError(
+          GSD_GIT_ERROR,
+          `Checked out branch "${branch}" successfully, but stash restore failed. ` +
+          `Your working tree changes are preserved in the git stash. ` +
+          `Run \`git stash pop\` to restore them manually. Cause: ${msg}`,
+        );
+      }
+    }
   } catch (err) {
-    if (stashed) {
+    // Only reach here when checkout itself failed (checkoutSucceeded is false).
+    // The stash-pop-after-success error path throws directly above and does
+    // not fall through to this catch.
+    if (stashed && !checkoutSucceeded) {
       try {
         popStashByRef(basePath, stashMarker);
       } catch (restoreErr) {
         logWarning("worktree", `git stash pop failed during checkout restore: ${restoreErr instanceof Error ? restoreErr.message : String(restoreErr)}`);
       }
     }
     throw err;
   }
 }
```
</requirement> -->

---

Failed to handle agent chat message. Please try again.



<!-- This is an auto-generated reply by CodeRabbit -->

Approved status-check tasks:
None

Approved documentation tasks:
None

When done:
1) Run the relevant verification for your changes.
2) Leave any changed files in the worktree for the babysitter app to finalize.
3) For each feedback item you addressed or were blocked on, emit a summary block in the following format:
   FEEDBACK_SUMMARY_START <auditToken>
   <A concise 1-2 sentence summary of what you did or why you were blocked>
   FEEDBACK_SUMMARY_END
   Include one block per audit token. These summaries will be posted as follow-up comments on the PR.
4) If documentation tasks were assigned, emit exactly one docs summary block in the following format:
   DOCS_SUMMARY_START <changed|no_change>
   <A concise 1-2 sentence summary of the docs you updated, or why no docs changes were necessary after inspection>
   DOCS_SUMMARY_END
````

</details>

File: src/resources/extensions/gsd/tests/checkout-branch-stash-guard.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/tests/checkout-branch-stash-guard.test.ts:63]: Addressed in commit `97ad60c` by the latest babysitter run.

File: src/resources/extensions/gsd/auto-worktree.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/auto-worktree.ts]: Addressed in commit `97ad60c` by the latest babysitter run.
